catalog: use correct ctx when releasing lease

andreimatei opened the pull request:
A recent refactoring used the wrong ctx when calling releaseLease(). A
different ctx needs to be used in that singleflight async code region,
as a comment explains, in order to not inherit the caller's
cancellation. Additionally, using the wrong ctx can also result in a
span use-after-Finish as the caller doesn't always wait for the
singleflight.

Fixes https://github.com/cockroachdb/cockroach/issues/76926
Fixes https://github.com/cockroachdb/cockroach/issues/77031
Fixes https://github.com/cockroachdb/cockroach/issues/76650

Release note: None
Release justification: bug fix

andreimatei committed:
catalog: use correct ctx when releasing lease

A recent refactoring used the wrong ctx when calling releaseLease(). A
different ctx needs to be used in that singleflight async code region,
as a comment explains, in order to not inherit the caller's
cancellation. Additionally, using the wrong ctx can also result in a
span use-after-Finish as the caller doesn't always wait for the
singleflight.

Fixes #76926
Fixes #77031
Fixes #76650

Release note: None
Release justification: bug fix

andreimatei committed:
catalog: small cleanup - shadow a ctx

This patch avoids having two (or actually three) ctx's in scope at the
same time by making the second shadow the first. Having two contexts is
error-prone as it's not clear which one to use and the casual
refactorer can easily get it wrong (as the previous commit fixes).

Release note: None
Release justification: minor cleanup

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/77163)
<!-- Reviewable:end -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Much appreciated 

> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77163#-:-Mx0uvoLBM29u8juHg3p:bpjjrfc)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz)



<!-- Sent from Reviewable.io -->

ajwerner commented:
Apparently every commit needs the justification?

ajwerner commented:
bors r+

ajwerner commented:
bors r-

andreimatei committed:
lint: allowing shadowing for variables named "ctx"

This patch adds an exception to the no-shadowing linter for variables
named "ctx", similar to the exception we have for variables named "err".
It's customary to name context.Context's ctx, to the point where being
forced to come up with a different name to avoid shadowing is harmful,
as it's too easy to use the wrong variable when refactoring (as another
commit in this PR exemplifies).

Release note: None
Release justification: non-production code change

andreimatei commented:
bors r+
